flink-connector/src/main/java/org/apache/flink/connector/delta/sink/committer/DeltaGlobalCommitter.java

```diff
@@ -222,6 +242,57 @@ private String resolveAppId(List<DeltaGlobalCommittable> globalCommittables) {
             .collect(Collectors.toList());
     }
 
+    /**
+     * Resolves whether to add the {@link Metadata} object to the transaction.
+     *
+     * <p>During this process:
+     * <ol>
+     *   <li>first we prepare metadata {@link Action} object using provided {@link RowType} (and
+     *       converting it to {@link StructType}) and partition values,
+     *   <li>then we compare the schema from above metadata with the current table's schema,
+     *   <li>resolved metadata object is added to the transaction only when it's the first commit to
+     *       the given Delta table or when the schemas are not matching but the sink was provided
+     *       with option {@link DeltaGlobalCommitter#canTryUpdateSchema} set to true (the commit
+     *       may still fail though if the Delta Standalone Writer will determine that the schemas
+     *       are not compatible),
+     *   <li>if the schemas are not matching and {@link DeltaGlobalCommitter#canTryUpdateSchema}
+     *       was set to false then we throw an exception
+     *   <li>if the schemas are matching then we do nothing and let the transaction proceed
+     * </ol>
+     * <p>
+     *
+     * @param currentTableVersion current version of the table's snapshot
+     * @param transaction         DeltaLog's transaction object
+     */
+    private void handleMetadataUpdate(long currentTableVersion,
+                                      OptimisticTransaction transaction) {
+        Metadata metadataAction = prepareMetadata();
+        StructType currentTableSchema = transaction.metadata().getSchema();
+        StructType commitSchema = metadataAction.getSchema();
+        boolean schemasAreMatching = areSchemasEqual(currentTableSchema, commitSchema);
+        if ((currentTableVersion == -1) || (!schemasAreMatching && canTryUpdateSchema)) {
+            transaction.updateMetadata(metadataAction);
+        } else if (!schemasAreMatching) {
+            String printableCurrentTableSchema = currentTableSchema == null ? "null"
+                : currentTableSchema.toPrettyJson();
+            String printableCommitSchema = commitSchema == null ? "null"
+                : commitSchema.toPrettyJson();
+
+            throw new RuntimeException(
+                "DataStream's schema is different from current table's schema. \n" +
+                    "provided: " + printableCurrentTableSchema + "\n" +
+                    "is different from: " + printableCommitSchema);
```

Comment:
i think the diffing can be made much smarter. here is the code in delta-spark to do that - https://github.com/delta-io/delta/blob/master/core/src/main/scala/org/apache/spark/sql/delta/schema/SchemaUtils.scala#L265 

we dont have to do this in this PR... but this is something that can be added as API in the standalone project and simply used here. @scottsand-db mind creating an issue about this, so that we dont forget.